src/gt4py/testing/suites.py

```diff
@@ -465,114 +466,99 @@ def _test_generation(self, test, externals_dict):
-                )
+        domain = shape - (Index(max_boundary.lower_indices) + Index(max_boundary.upper_indices))
 
-        new_boundary = tuple(
-            (max(abs(b[0]), abs(mb[0])), max(abs(b[1]), abs(mb[1])))
-            for b, mb in zip(cls.max_boundary, max_boundary)
+        referenced_inputs = {
+            name: info for name, info in implementation.field_info.items() if info is not None
+        }
+        referenced_inputs.update(
+            {name: info for name, info in implementation.parameter_info.items() if info is not None}
         )
 
-        shape = None
-        for field in fields.values():
-            if isinstance(field, np.ndarray):
-                assert field.shape == (shape if shape is not None else field.shape)
-                shape = field.shape
-
-        patched_origin = tuple(nb[0] for nb in new_boundary)
-        patching_origin = tuple(po - o for po, o in zip(patched_origin, origin))
-        patched_shape = tuple(nb[0] + nb[1] + d for nb, d in zip(new_boundary, domain))
-        patching_slices = [slice(po, po + s) for po, s in zip(patching_origin, shape)]
-
+        # set externals for validation method
         for k, v in implementation.constants.items():
             sys.modules[self.__module__].__dict__[k] = v
 
+        # copy input data to storages
         inputs = {}
-        for k, f in fields.items():
-            if isinstance(f, np.ndarray):
-                patched_f = np.empty(shape=patched_shape)
-                patched_f[patching_slices] = f
-                inputs[k] = gt_storage.from_array(
-                    patched_f,
-                    dtype=f.dtype,
-                    shape=patched_f.shape,
-                    default_origin=patched_origin,
-                    backend=implementation.backend,
+        info: FieldInfo
+        for name, data in input_data.items():
+            data = input_data[name]
+            if name in referenced_inputs:
+                info = referenced_inputs[name]
+                if isinstance(info, FieldInfo):
+                    inputs[name] = gt_storage.from_array(
+                        data,
+                        dtype=data.dtype,
+                        shape=shape,
+                        default_origin=origin,
+                        backend=implementation.backend,
+                    )
+                else:
+                    inputs[name] = data
+            else:
+                inputs[name] = None
+
+        # copy input data for validation
+        validation_inputs = {}
+        info: FieldInfo
+        for name, data in input_data.items():
+            data = input_data[name]
+            if name in referenced_inputs:
+                info = referenced_inputs[name]
+                if isinstance(info, FieldInfo):
+                    validation_inputs[name] = np.array(data)
+                else:
+                    validation_inputs[name] = data
+            else:
+                validation_inputs[name] = None
```

Comment:
These two loops are basically doing the same for input storages and validation arrays, right? I think it would be easier to merge them together.

Reply:
Done